src/backend/distributed/utils/resource_lock.c

```diff
@@ -728,7 +725,10 @@ SerializeNonCommutativeWrites(List *shardIntervalList, LOCKMODE lockMode)
 			LockReferencedReferenceShardResources(firstShardInterval->shardId, lockMode);
 		}
 	}
-
+	if (anyTableReplicated && ClusterHasKnownMetadataWorkers() && !IsFirstWorkerNode())
```

Comment:
note that with this change we are diverging the behavior of Postgres by first locking the referenced tables/shards when there is a foreign key. Postgres acquires the locks on main table first, then cascades into the referenced tables.

So could this change might introduce new classes of concurrency issues where there are concurrent modifications to the `ReferencedTables` (e.g., distributed table) while there are modifications to the main table (e.g., reference table)?

Also this code doesn't look like what the PR description tells, it seems doing the opposite. 

It feels like the code should be much more explicit about what we are doing, for example, the following code block is how I think this logic should look like -- though probably the code could be nicer:

```C


bool acquiredLocally = false

// simplest case, no query from any node
// first acquire the locks on the modified table
// then cascade into the referenced tables
if (!ClusterHasKnownMetadataWorkers())
	LockShardListResources(shardIntervalList, lockMode);
	for replicatedTable in anyTableReplicated
		LockReferencedReferenceShardResourcesLocally(replicatedTable->firstShardInterval->shardId, lockMode);

	acquiredLocally = true
else
	
	// moderate case, we are the first worker node
	// acquire the locks locally, same as the simplest case
	if IsFirstWorkerNode() && anyTableReplicated
		LockShardListResources(shardIntervalList, lockMode);
		for replicatedShard in replicatedShards
			LockReferencedReferenceShardResourcesLocally(replicatedTable->firstShardInterval->shardId, lockMode);
		acquiredLocally = true

	// ok, not we are on a worker that is not first worker node or we are on the coordinator
	else
		LockShardListResourcesOnFirstWorker(lockMode, replicatedShardList);
		for replicatedShard in replicatedShards
			LockReferencedReferenceShardResourcesOnTheFirstNode(..)

// ok, above we serialized all the modifications on the relevant tables across nodes
// by acquiring the locks on the first worker node
// so no two concurrent modifications across multiple nodes could interleave any modifications

// so, now lets always acquire the locks locally as there are several other concurrency scenarios
// where we expect shard resource locks are acquired locally
if !acquiredLocally 
	LockShardListResources(shardIntervalList, lockMode);
	for replicatedShard in replicatedShards
		LockReferencedReferenceShardResourcesLocally(replicatedTable->firstShardInterval->shardId, lockMode);
```

note that I think it is even OK to drop all code relevant to `!ClusterHasKnownMetadataWorkers()` given that with Citus 11.0 we expect all clusters to have metadata synced. This is like a safe-guard in case old clusters upgraded to CItus 11.0 could not sync the metadata.

Reply:
Thanks for the lead regarding the lock order. 

I still think the PR description is right. But you are right that the commit description is wrong. Please let me know if I miss anything. Prior to the changes, "`LockShardListResources(shardIntervalList, lockMode);`" was the last statement, which acquires the locks for the modified table locally.  It is also the last lock acquired when the request is initiated from the first worker node.  In particular, the PR changes the order in which the locks are acquired on the first worker node with respect to the node that received the request. 
The first version acquired locks on the reference tables and then the modified table. 

As you suggested, I changed the order.  With the last commit, independent of the node that received the request, the locks are acquired for the modified table and then the reference tables on the first node. 

I noticed your suggestion to acquire remote locks before local ones. In the current version, locks are acquired for the modified table on the first worker, followed by local locks. Similarly, locks for reference tables are obtained on the first worker and then locally.  I haven’t incorporated that suggestion yet.